File: py/week20211205-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getDirections(self, root: Optional[TreeNode], startValue: int, destValue: int) -> str:
        startNode = None
        cnt = 0


        def dfs(node, p):
            nonlocal startNode, cnt
            if node:
                node.parent = p
                cnt+=1
                dfs(node.left, node)
                dfs(node.right, node)
                if node.val == startValue:
                    startNode = node

        dfs(root, None)

        visited = [False for _ in range(cnt + 1)]
        ans = None

        def dfsPath(node, prev:str):
            nonlocal ans
            if node.val == destValue:
                ans = prev
                return 

            if ans is not None:
                return

            visited[node.val] = True

            if node.left and not visited[node.left.val]:
                dfsPath(node.left, prev + 'L')

            if node.right and not visited[node.right.val]:
                dfsPath(node.right, prev + 'R')

            if node.parent and not visited[node.parent.val]:
                dfsPath(node.parent, prev + 'U')

        dfsPath(startNode, '')

        return ans

File: java/week20211205-3.java
import java.util.*;

import javax.swing.tree.TreeNode;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    class Step {
        public int cnt;
        public String path;
        public TreeNode cur;

        public Step(String path, TreeNode cur) {
            this.path = path;
            this.cur = cur;
            this.cnt = path.length();
        }
    }


    public String getDirections(TreeNode root, int startValue, int destValue) {
        parent = new HashMap<>();
        dest = destValue;
        start = startValue;
        
        dfs(root, null);

        int n = parent.size();
        visited = new boolean[n + 1];

        Queue<Step> queue = new PriorityQueue<>((s1, s2) -> s1.cnt - s2.cnt);
        queue.add(new Step("", startNode));

        while (!queue.isEmpty()) {
            Step s = queue.poll();
            TreeNode node = s.cur;

            if (node.val == destValue)
                return s.path;

            visited[node.val] = true;

            if (node.left != null && !visited[node.left.val]) {
                queue.add(new Step(s.path + "L", node.left));
            }

            if (node.right != null && !visited[node.right.val]) {
                queue.add(new Step(s.path + "R", node.right));
            }

            TreeNode p = parent.get(node.val);
            if (p != null && !visited[p.val]) {
                queue.add(new Step(s.path + "U", p));
            }
        }

        // dfsPath(startNode, "");

        return ans;
    }

    public void dfsPath(TreeNode node, String prevPath) {
        if (node.val == dest) {
            ans = prevPath;
            return;
        }

        if (ans != null)
            return;

        System.out.printf("%d %s\n", node.val, prevPath);

        visited[node.val] = true;

        if (node.left != null && !visited[node.left.val]) {
            dfsPath(node.left, prevPath + "L");
        }

        if (node.right != null && !visited[node.right.val]) {
            dfsPath(node.right, prevPath + "R");
        }

        TreeNode p = parent.get(node.val);
        if (p != null && !visited[p.val]) {
            dfsPath(p, prevPath + "U");
        }
    }

    boolean[] visited;
    int dest;
    int start;
    String ans;
    TreeNode startNode;
    Map<Integer, TreeNode> parent;

    public void dfs(TreeNode node, TreeNode p) {
        if (node != null) {
            parent.put(node.val, p);
            dfs(node.left, node);
            dfs(node.right, node);

            if (node.val == start)
                startNode = node;
        }
    }
}